diff --git a/stack/btm/btm_dev.cc b/stack/btm/btm_dev.cc
index 5368fad..d700e93 100644
--- a/stack/btm/btm_dev.cc
+++ b/stack/btm/btm_dev.cc
@@ -136,12 +136,12 @@ bool BTM_SecAddDevice(const RawAddress& bd_addr, DEV_CLASS dev_class,
     }
   }
 
-#if (BTIF_MIXED_MODE_INCLUDED == TRUE)
+//#if (BTIF_MIXED_MODE_INCLUDED == TRUE)
   if (key_type < BTM_MAX_PRE_SM4_LKEY_TYPE)
     p_dev_rec->sm4 = BTM_SM4_KNOWN;
   else
     p_dev_rec->sm4 = BTM_SM4_TRUE;
-#endif
+//#endif
 
   p_dev_rec->rmt_io_caps = io_cap;
   p_dev_rec->device_type |= BT_DEVICE_TYPE_BREDR;
