diff --git a/drivers/mmc/core/core.c b/drivers/mmc/core/core.c
index 3e3c79f..00a8a3c 100644
--- a/drivers/mmc/core/core.c
+++ b/drivers/mmc/core/core.c
@@ -160,19 +160,22 @@ void mmc_request_done(struct mmc_host *host, struct mmc_request *mrq)
 		led_trigger_event(host->led, LED_OFF);
 
 		if (mrq->sbc) {
-			pr_debug("%s: req done <CMD%u>: %d: %08x %08x %08x %08x\n",
+		if ((host->restrict_caps & RESTRICT_CARD_TYPE_SDIO))
+			pr_err("%s: req done <CMD%u>: %d: %08x %08x %08x %08x\n",
 				mmc_hostname(host), mrq->sbc->opcode,
 				mrq->sbc->error,
 				mrq->sbc->resp[0], mrq->sbc->resp[1],
 				mrq->sbc->resp[2], mrq->sbc->resp[3]);
 		}
 
+		if ((host->restrict_caps & RESTRICT_CARD_TYPE_SDIO))
 		pr_debug("%s: req done (CMD%u): %d: %08x %08x %08x %08x\n",
 			mmc_hostname(host), cmd->opcode, err,
 			cmd->resp[0], cmd->resp[1],
 			cmd->resp[2], cmd->resp[3]);
 
 		if (mrq->data) {
+		if ((host->restrict_caps & RESTRICT_CARD_TYPE_SDIO))
 			pr_debug("%s:     %d bytes transferred: %d\n",
 				mmc_hostname(host),
 				mrq->data->bytes_xfered, mrq->data->error);
@@ -194,6 +197,7 @@ void mmc_request_done(struct mmc_host *host, struct mmc_request *mrq)
 		}
 
 		if (mrq->stop) {
+		if ((host->restrict_caps & RESTRICT_CARD_TYPE_SDIO))
 			pr_debug("%s:     (CMD%u): %d: %08x %08x %08x %08x\n",
 				mmc_hostname(host), mrq->stop->opcode,
 				mrq->stop->error,
@@ -252,16 +256,19 @@ static int mmc_start_request(struct mmc_host *host, struct mmc_request *mrq)
 		return -ENOMEDIUM;
 
 	if (mrq->sbc) {
+		if ((host->restrict_caps & RESTRICT_CARD_TYPE_SDIO))
 		pr_debug("<%s: starting CMD%u arg %08x flags %08x>\n",
 			 mmc_hostname(host), mrq->sbc->opcode,
 			 mrq->sbc->arg, mrq->sbc->flags);
 	}
 
+	if ((host->restrict_caps & RESTRICT_CARD_TYPE_SDIO))
 	pr_debug("%s: starting CMD%u arg %08x flags %08x\n",
 		 mmc_hostname(host), mrq->cmd->opcode,
 		 mrq->cmd->arg, mrq->cmd->flags);
 
 	if (mrq->data) {
+		if ((host->restrict_caps & RESTRICT_CARD_TYPE_SDIO))
 		pr_debug("%s:     blksz %d blocks %d flags %08x "
 			"tsac %d ms nsac %d\n",
 			mmc_hostname(host), mrq->data->blksz,
@@ -271,6 +278,7 @@ static int mmc_start_request(struct mmc_host *host, struct mmc_request *mrq)
 	}
 
 	if (mrq->stop) {
+		if ((host->restrict_caps & RESTRICT_CARD_TYPE_SDIO))
 		pr_debug("%s:     CMD%u arg %08x flags %08x\n",
 			 mmc_hostname(host), mrq->stop->opcode,
 			 mrq->stop->arg, mrq->stop->flags);
@@ -531,6 +539,7 @@ static void mmc_wait_for_req_done(struct mmc_host *host,
 
 		mmc_retune_recheck(host);
 
+		if ((host->restrict_caps & RESTRICT_CARD_TYPE_SDIO))
 		pr_debug("%s: req failed (CMD%u): %d, retrying...\n",
 			 mmc_hostname(host), cmd->opcode, cmd->error);
 		cmd->retries--;
