识别不到卡，主要是pid vid没有加进去驱动里（当然最好是提供一个支持的驱动，可以尝试修改，如果不行那也没有办法找模组原厂拿新的驱动）

目前我加进去了有东西，但是
[  651.358066] RTL871X: HalPwrSeqCmdParsing: Fail to polling Offset[0x5]=01
这个是usb供电的问题，所以有些模块就是不一样，它就是不行，要它能用就要改电路或者各种


--- a/drivers/net/wireless/rockchip_wlan/rtl8812au/os_dep/linux/usb_intf.c
+++ b/drivers/net/wireless/rockchip_wlan/rtl8812au/os_dep/linux/usb_intf.c
@@ -283,6 +283,7 @@ static struct usb_device_id rtw_usb_id_tbl[] ={
        {USB_DEVICE(0x050D, 0x1106),.driver_info = RTL8812}, /* Belkin - sercomm */
        {USB_DEVICE(0x2001, 0x330E),.driver_info = RTL8812}, /* D-Link - ALPHA */
        {USB_DEVICE(0x7392, 0xA822),.driver_info = RTL8812}, /* Edimax - Edimax */
+       {USB_DEVICE(0x0411, 0x029B),.driver_info = RTL8812}, /* BUFFALP - BUFFALO */
        {USB_DEVICE(0x0DF6, 0x0074),.driver_info = RTL8812}, /* Sitecom - Edimax */
        {USB_DEVICE(0x04BB, 0x0952),.driver_info = RTL8812}, /* I-O DATA - Edimax */
        {USB_DEVICE(0x0789, 0x016E),.driver_info = RTL8812}, /* Logitec - Edimax */
